spec.html

```diff
@@ -23067,7 +23283,381 @@ <h1>Example Cyclic Module Record Graphs</h1>
+          <p>Alternatively, consider a case where _A_ has an evaluation error; for example, its source code throws an exception. In that case, the evaluation-time analog of the above steps still occurs, including the early return from the second call to InnerModuleEvaluation on _A_. However, once we unwind back to the original InnerModuleEvaluation on _A_, it fails by assumption. The exception thrown propagates up to _A_.Evaluate(), which records the error in all modules that are currently on its _stack_ (i.e., the modules that are still ~evaluating~) as well as via [[AsyncParentModules]], which form a chain for modules which contain or depend on top-level `await` through the whole dependency graph through the AsyncModuleExecutionRejected algorithm. Hence both _A_ and _B_ become ~evaluated~ and the exception is recorded in both _A_ and _B_'s [[EvaluationError]] fields, while _C_ is left as ~evaluated~ with no [[EvaluationError]].</p>
+
+          <p>Lastly, consider a module graph with a cycle, where all modules complete asynchronously:</p>
+          <emu-figure id="figure-module-graph-cycle-async" caption="An asynchronous cyclic module graph">
+            <img alt="A module graph in which module A depends on module B and C, module B depends on module D, module C depends on module D and E, and module D depends on module A" width="241" height="211" src="img/module-graph-cycle-async.svg">
+          </emu-figure>
+          <p>Linking happens as before, and all modules end up with [[Status]] set to ~linked~.</p>
+
+          <p>Calling _A_.Evaluate() triggers InnerModuleEvaluation (calls `.then()`) on _A_, _B_, and _D_, which all transition to ~evaluating~. Then InnerModuleEvaluation is called on _A_ again, which is a no-op because it is already ~evaluating~. At this point, _D_.[[PendingAsyncDependencies]] is 0, so ExecuteAsyncModule(_D_) is called and triggers the execution promise for _D_. We unwind back to the InnerModuleEvaluation on _B_, setting _B_.[[PendingAsyncDependencies]] to 1 and _B_.[[AsyncEvaluation]] to *true*. We unwind back to the original InnerModuleEvaluation on _A_, setting _A_.[[PendingAsyncDependencies]] to 1. In the next iteration of the loop over _A_'s dependencies, we call InnerModuleEvaluation on _C_ and thus on _D_ (again a no-op) and _E_. As _E_ has no dependencies, ExecuteAsyncModule(_E_) is called, which triggers its execution promise. Because _E_ is not part of a cycle, it is immediately removed from the stack and transitions to ~evaluating-async~. We unwind once more to the original InnerModuleEvaluation on _A_, setting _C_.[[AsyncEvaluation]] to *true*. Now we finish the loop over _A_'s dependencies, set _A_.[[AsyncEvaluation]] to *true*, and remove the entire strongly connected component from the stack, transitioning all of the modules to ~evaluating-async~ at once. At this point, the fields of the modules are as given in <emu-xref href="#table-module-graph-cycle-async-fields-1"></emu-xref>.</p>
```

Comment:
There's still instances of "triggers the execution promise of", which I'd like to be spelled out more explicitly.

In particular, what does "As _E_ has no dependencies, ExecuteAsyncModule(_E_) is called, which triggers its execution promise." mean precisely? It doesn't seem to mean that _E_ finished evaluating and fulfilled its promise, since the next paragraph talks about _E_ finishing executing first.

Reply:
Hopefully this is clearer now

Reply:
You should delete one of the two spaces before "We unwind". (It's causing the 'ecmarkup' run to fail.)